Addition of WiFiStream library and standard-firmata-wifi.ino sketch

turkycat opened the pull request.

turkycat committed:
Adding WiFiStream class to /lib/

Well tested with OPEN networks using both DHCP and static IP configurations.

turkycat committed:
Added WEP encryption support.

Also removed 10sec delay after WiFi.begin() as it is unnecessary despite the sample sketches using this behavior. By the time WiFi.begin() returns a result, the connection is either established or failed with error.

turkycat committed:
Added WPA encryption support.

Also added a config() function which provides another way to set static IPs before calling begin().

turkycat committed:
Adding standard-firmata-wifi.ino sketch with support for all encryption types.

turkycat committed:
Updated documentation.

Made necessary corrections to reflect the repository's name change.
Also modified the ethernet sketch to no longer use the #define DHCP as I felt it was unncessary when using #define STATIC_IP_ADDRESS. Documentation updated to reflect this change.

peterschwarz commented:
This looks great!  What WiFi board did you test it with?  Were you also able to test it with an ethernet sheild?

turkycat commented:
I tested with the Arduino WiFi shield. I also have an ethernet shield, and I did test your sketch before and after I made the changes to it. I made network connections to the Arduino from several devices and successfully blinked an LED with each shield/sketch combo. In the case of WiFi, I tested each security type as well using a Linksys router.

peterschwarz commented:
Very cool.  I'll still test it out - my UNO is screwed onto a robot chassis, so I'll need to figure out how to wire up a micro to my ethernet shield.  

For wi-fi, I have this: https://www.adafruit.com/products/1469, which may or may not be usable.  I won't test it, but I might keep it in mind when looking through your code.

turkycat commented:
Thanks for the link. I dissected the WiFi library when I was building up the solution, and I can say from first glance that the Huzzah board won't work. The SPI pins are all identical, but unfortunately the WiFi library also uses pin 7 as a handshaking pin, and the Huzzah board does not. The WiFi library won't work without this pin, and it looks like Adafruit provides their own board-specific library for this device: https://learn.adafruit.com/adafruit-cc3000-wifi/cc3000-library-software

I may be interested in adding support for this board, but it would likely require an entirely new class as it would be a waste of memory to include both libraries in one. We'd also have to add a dependency to the library's GH repo: https://github.com/adafruit/Adafruit_CC3000_Library/

peterschwarz commented:
Yeah, kind of like this: [standard-firmata-cc3000](https://github.com/peterschwarz/standard-firmata-cc3000). Though, I had a rough go of it (I think it might say as much in the README).  It might be a starting point to merging the libraries in a way.

turkycat commented:
ah ;)
I'll have to get myself a cc3000 first, but I may actually do this to see if we can get a wider range of shield support.

peterschwarz commented:
How did you build this?  I'm having trouble getting it to build via the build script.  

On linux, I get the following:

```
-------------------------
Arduino.mk Configuration:
- [AUTODETECTED]       CURRENT_OS = LINUX 
- [USER]               ARDMK_DIR = ./Arduino-Makefile 
- [USER]               ARDUINO_DIR = /usr/share/arduino 
- [AUTODETECTED]       ARDUINO_VERSION = 105 
- [DEFAULT]            ARDUINO_SKETCHBOOK = /home/pschwarz/sketchbook 
- [USER]               AVR_TOOLS_DIR = /usr/bin 
- [COMPUTED]           ARDUINO_LIB_PATH = /usr/share/arduino/libraries (from ARDUINO_DIR)
- [DEFAULT]            ARDUINO_CORE_PATH = /usr/share/arduino/hardware/arduino/cores/arduino 
- [COMPUTED]           ARDUINO_VAR_PATH = /usr/share/arduino/hardware/arduino/variants (from ARDUINO_DIR)
- [COMPUTED]           BOARDS_TXT = /usr/share/arduino/hardware/arduino/boards.txt (from ARDUINO_DIR)
- [USER]               USER_LIB_PATH = ./lib 
- [DEFAULT]            PRE_BUILD_HOOK = pre-build-hook.sh 
- [USER]               BOARD_TAG = uno 
- [USER]               OBJDIR = ./bin/uno/standard-firmata-networking 
Arduino-Makefile/Arduino.mk:666: *** Need exactly one .pde or .ino file. This makefile doesn't support multiple .ino/.pde files yet.  Stop.
```

turkycat commented:
Hmm, I'm using Windows and the Arduino IDE to upload sketches. I'm fairly familiar with linux, but don't have the setup necessary to test on that platform. It looks to me like it is complaining that two *.ino files are in the build directory. Perhaps restructuring the location of the .ino files will resolve the issue?

Is this script building intermediate binaries? I'm not sure if it is the same as a typical Makefile, but you may be able to build dependency chains where the wifi and ethernet .ino files will not depend on each other and will build separate binaries. I'm open to any suggestions

turkycat committed:
Removing DHCP section of code that was commented out in wifi sketch.

Also moved wifi sketch to \src\ directory

peterschwarz commented:
I've converted the build tool to use [PlatformIO](http://platformio.org/) via #2

See if you can get things building using this.

turkycat commented:
haven't rebased my work yet, but I got a successful build using platformio. However, I wasn't able to upload to my Arduino because platformio cannot seem to bind to my com ports successfully, gives out of sync error.
In either case, I'd like to add another section below your default 'installation' section to cater to those more familiar with Windows and the Arduino IDE, as installing additional software (python and platformio) shouldn't be necessary in my opinion. I'd like both options at least, thoughts?

turkycat commented:
I retract the part about failed com port attachment, I was able to fix this with a reboot.

turkycat committed:
Merge branch 'WiFi'

turkycat commented:
I rebased my work on your platformio changes, corrected documentation conflicts with THEIRS strategy, added a windows setup option (far easier than using platformio, for Windows users at least).

I attempted platformio build again, but am getting errors for function definitions (it is trying to compile both sketches together). Are you really familiar with this build system? Any idea how to segregate these better or specify a single build target?

waltonryan commented:
Is it now possible to use this firmata with the cc3000 wifi? I'm looking to get wifi capabilities with the johnny-five library.

peterschwarz commented:
@waltonryan You could use this as a starting point: [standard-firmata-cc3000](https://github.com/peterschwarz/standard-firmata-cc3000).  I didn't have luck with it (and got busy with paid work), but it's close I think.

waltonryan commented:
Cool, thanks, I'll take a look.

mares0909 commented:
"I'm looking to get wifi capabilities with the johnny-five library." @waltonryan 
That's exactly what I'm looking for as well!
I have a NodeMCU and a few ESP8266 modules. They can be programmed with the Arduino IDE and as far as I'm aware most of the Wi-Fi features are inherited from the original CC3300 library. If standard-firmata-networking could be used between the ESP8266 and Johnny-Five that would be pretty cool!

waltonryan commented:
It would be cool. From what I remember when I was looking into it, the standard-firmata-cc3000 was a bit cumbersome to get it going.

nonameolsson commented:
What's the status of this?
